File: leecode/14.py
#!/user/bin/python
#-*- coding:utf-8 -*-

class Solution:

    # T(n) = O(len(strs) * len(strs[0]))
    # S(n) = O(len(str[0]))
    def longestCommonPrefix1(self, strs):
        """
        :type strs: List[str]
        :rtype: str
        """
        if len(strs) == 0:
            return ''
        if len(strs) == 1:
            return strs[0]

        result = []
        for i in range(len(strs[0])):
            c = strs[0][i]
            if all(map(lambda str: False if (len(str) < (i + 1)) else (str[i] == c), strs)):
                result.append(c)
            else:
                break
        return ''.join(result)

    

    #divide and conquer
    def longestCommonPrefix(self, strs):
        """
        :type strs: List[str]
        :rtype: str
        """
        if len(strs) == 0:
            return ''
        if len(strs) == 1:
            return strs[0]

        length = len(strs)
        comPrefix1 = self.longestCommonPrefix(strs[0 : int(length/2)])
        comPrefix2 = self.longestCommonPrefix(strs[int(length/2) : ])
        len1 = len(comPrefix1)
        len2 = len(comPrefix2)
        # if one is empty, return ''
        if 0 == len1 or 0 == len2:
            return ''
        # make comPrefix1 samller
        if len1 > len2:
            comPrefix1, comPrefix2 = comPrefix2, comPrefix1
        # get the common prefix
        result = []
        for i in range(len(comPrefix1)):
            if comPrefix1[i] == comPrefix2[i]:
                result.append(comPrefix1[i])
            else:
                break
        return ''.join(result)        





import unittest
class SolutionTestCase(unittest.TestCase):
    def testMatch(self):
        solution = Solution()
        self.assertTrue(solution.longestCommonPrefix(["flower","flow","flight"]) == "fl")
        self.assertTrue(solution.longestCommonPrefix(["dog","racecar","car"]) == "")

if __name__ == "__main__": unittest.main()

File: leecode/14.cpp
#include <assert.h>
#include <string.h>
#include <stdlib.h>

int MAX_SIZE = 1024;
char buf[1024];
int pos = 0;

char* alloc(int size) {
    if (pos + size < MAX_SIZE) {
        int temp = pos;
        pos += size;
        return buf + temp;
    }
    return NULL;
}

// 递归中如果碰到要分配内存非常麻烦，遇到这种情况有时alloc可以帮上忙
char* longestCommonPrefix(char** strs, int strsSize) {
    if (0 == strsSize) {
        return "";
    }
    if (1 == strsSize) {
        return strs[0];
    }

    char* comPrefix1 = longestCommonPrefix(strs, strsSize/2);
    char* comPrefix2 = longestCommonPrefix(strs + strsSize/2, strsSize - strsSize/2);
    
    int len1 = strlen(comPrefix1);
    int len2 = strlen(comPrefix2);
    // if one of comPrefix is empty, return ""
    if (len1 == 0 || len2 == 0) {
        return "";
    }

    // get the small length of length
    int minLen = len1 > len2 ? len1 : len2;
    // malloc space
    char * result = (char*)alloc(minLen + 1);

    // compare
    int i = 0;
    while (i < minLen) {
        if (comPrefix1[i] != comPrefix2[i]) {
            break;
        }
        result[i] = comPrefix1[i];
        i++;
    }

    // return result
    return result;
}



void main() {
    char* input1[3] = {"flower","flow","flight"};
    assert(!strcmp(longestCommonPrefix(input1, 3), "fl"));
    char* input2[3] = {"dog","racecar","car"};
    assert(!strcmp(longestCommonPrefix(input2, 3), ""));
}